### PR TITLE
ComplexTable with sorted buckets

### PR DESCRIPTION
**Pull request by hillmich**
- Insertion in sorted order
- Lookup with early breaks based on sorted order
- Adjusted tests

### CODE REVIEW

File: include/dd/ComplexTable.hpp [hunk 1]
```diff
@@ -135,70 +144,45 @@ namespace dd {
...
-            }
-
             // search in (potentially) lower bucket
             if (val - TOLERANCE >= 0) {
                 const auto lowerKey = hash(val - TOLERANCE);
                 if (lowerKey != key) {
-                    const auto& lowerBucket = table[lowerKey];
-                    p                       = find(lowerBucket, val);
+                    ++neighbors;
+                    Entry* p = find(table[lowerKey], val);
```
Comment (burgholzer) [inline include/dd/ComplexTable.hpp:167]:
```suggestion
                    auto p = find(table[lowerKey], val);
```
consider using `auto` for staying consistent with the above statement

File: include/dd/ComplexTable.hpp [hunk 2]
```diff
@@ -135,70 +144,45 @@ namespace dd {
...
             }
 
             // search in (potentially) higher bucket
-            const auto upperKey = hash(val - TOLERANCE);
+            const auto upperKey = hash(val + TOLERANCE);
             if (upperKey != key) {
-                const auto& upperBucket = table[upperKey];
-                p                       = find(upperBucket, val);
+                ++neighbors;
+                Entry* p = find(table[upperKey], val);
```
Comment (burgholzer) [inline include/dd/ComplexTable.hpp:178]:
```suggestion
               auto p = find(table[upperKey], val);
```
consider using `auto` for staying consistent with the above statement

File: include/dd/ComplexTable.hpp [hunk 3]
```diff
@@ -371,50 +362,98 @@ namespace dd {
...
+        std::size_t collisions = 0;
+        std::size_t hits = 0;
+        std::size_t lookups = 0;
+        std::size_t breaks = 0;
+        std::size_t neighbors = 0;
+
     private:
-        using Bucket = Entry*;
-        using Table  = std::array<Bucket, NBUCKET>;
+        using Bucket = Entry *;
```
Comment (burgholzer) [inline include/dd/ComplexTable.hpp:392]: does this really conform to clang-format? I'd expect it to be `Entry*`

File: include/dd/ComplexTable.hpp [hunk 4]
```diff
@@ -371,50 +362,98 @@ namespace dd {
...
 
         // numerical tolerance to be used for floating point values
         static inline fp TOLERANCE = 1e-13;
 
-        Entry*                                available{};
-        std::vector<std::vector<Entry>>       chunks{};
-        std::size_t                           chunkID;
+        Entry *available{};
+        std::vector<std::vector<Entry>> chunks{};
+        std::size_t chunkID;
```
Comment (burgholzer) [inline include/dd/ComplexTable.hpp:402]: this is not conforming to clang-format, is it?

File: include/dd/ComplexTable.hpp [hunk 5]
```diff
@@ -371,50 +362,98 @@ namespace dd {
...
+               << ", collisions: " << collisions
+               << ", looks: " << lookups
+               << ", breaks: " << breaks
+               << ", neighbors: " << neighbors
+               << ", hitRatio: " << hitRatio()
+               << ", colRatio: " << colRatio()
+               << ", gc calls: " << gcCalls
+               << ", gc runs: " << gcRuns
+               << "\n";
+            //clang-format on
```
Comment (burgholzer) [inline include/dd/ComplexTable.hpp:380]:
```suggestion
            // clang-format on
```
I guess clang format is not turned back on again when the whitespace is missing.

File: include/dd/ComplexTable.hpp [hunk 6]
```diff
@@ -371,50 +362,98 @@ namespace dd {
...
+
+        inline Entry *find(const Bucket &bucket, const fp &val) {
+            Entry *p = bucket;
             while (p != nullptr) {
                 if (std::abs(p->value - val) < TOLERANCE) {
                     ++hits;
                     return p;
                 }
+                if (val > p->value + TOLERANCE) {
+                    ++breaks;
```
Comment (burgholzer) [inline include/dd/ComplexTable.hpp:454]: is it on purpose that a `break` does not count as a `collision`?

File: include/dd/ComplexTable.hpp [hunk 7]
```diff
@@ -135,70 +144,45 @@ namespace dd {
...
-                } else {
-                    ++collisions;
-                }
-            } else if (key == sqrt2_2Key) {
-                if (std::abs(val - SQRT2_2) < TOLERANCE) {
-                    ++hits;
-                    return &sqrt2_2;
-                } else {
-                    ++collisions;
+            {
```
[the remark this answers is not in the document]
Comment (hillmich) [reply include/dd/ComplexTable.hpp:155]: Yeah, to limit the scope of p

File: include/dd/ComplexTable.hpp (hunk 1 above)
Comment (hillmich) [reply include/dd/ComplexTable.hpp:167]: Does not apply anymore.

File: include/dd/ComplexTable.hpp [hunk 8]
```diff
@@ -341,6 +327,10 @@ namespace dd {
             allocationSize = INITIAL_ALLOCATION_SIZE * GROWTH_FACTOR;
             allocations    = INITIAL_ALLOCATION_SIZE;
 
+            for (auto& entry: chunks[0]) {
+                entry.refCount = 0;
+            }
+
```
[the remark this answers is not in the document]
Comment (hillmich) [reply include/dd/ComplexTable.hpp:333]: I don't think that's gonna work to only set the `refCount`.

[the remark this answers is not in the document]
Comment (burgholzer) [reply include/dd/ComplexTable.hpp:343]: agreed. forget what I said 👍🏼

File: include/dd/ComplexTable.hpp [hunk 9]
```diff
@@ -133,31 +135,38 @@ namespace dd {
         Entry* lookup(const fp& val) {
             assert(!std::isnan(val));
             assert(val >= 0); // required anyway for the hash function
-
+            lookups++;
```
Comment (burgholzer) [inline include/dd/ComplexTable.hpp:138]: the comment right below this line should probably be removed ;o)

File: include/dd/ComplexTable.hpp [hunk 10]
```diff
@@ -166,17 +175,20 @@ namespace dd {
             // search in (potentially) lower bucket
             if (lowerKey != key) {
                 ++lower_neighbors;
-                Entry* p_lower = find(table[lowerKey], val);
-                if (p_lower != nullptr) {
+                // buckets are sorted so we only have to look into the last entry of the lower bucket
+                Entry* p_lower = last_entry_table[lowerKey];
+                if (p_lower != nullptr && val - p_lower->value < TOLERANCE) {
```
Comment (burgholzer) [inline include/dd/ComplexTable.hpp:180]:
```suggestion
                if (p_lower != nullptr && std::abs(val - p_lower->value) < TOLERANCE) {
```
shouldn't the absolute value be used here? or is it really guaranteed that `val >= p_lower->value`?

File: include/dd/ComplexTable.hpp [hunk 11]
```diff
@@ -166,17 +175,20 @@ namespace dd {
...
 
             // search in (potentially) higher bucket
             if (upperKey != key) {
                 ++upper_neighbors;
-                Entry* p_upper = find(table[upperKey], val);
-                if (p_upper != nullptr) {
+                // buckets are sorted, we only have to look at the first element
+
+                Entry* p_upper = table[upperKey];
+                if (p_upper != nullptr && p_upper->value - val < TOLERANCE) {
```
Comment (burgholzer) [inline include/dd/ComplexTable.hpp:191]: same as above

File: include/dd/ComplexTable.hpp [hunk 12]
```diff
@@ -334,9 +350,14 @@ namespace dd {
             count     = 0;
             peakCount = 0;
 
-            collisions = 0;
-            hits       = 0;
-            lookups    = 0;
+            collisions        = 0;
+            insert_collisions = 0;
```
Comment (burgholzer) [inline include/dd/ComplexTable.hpp:354]: I'd suggest to stick to one naming pattern throughout the project as much as possible.
consider using camelCase.
```suggestion
            insertCollisions = 0;
```

File: include/dd/ComplexTable.hpp [hunk 13]
```diff
@@ -334,9 +350,14 @@ namespace dd {
...
+            hits              = 0;
+            find_or_inserts   = 0;
```
Comment (burgholzer) [inline include/dd/ComplexTable.hpp:356]:
```suggestion
            findOrInserts   = 0;
```

File: include/dd/ComplexTable.hpp [hunk 14]
```diff
@@ -334,9 +350,14 @@ namespace dd {
...
+            lookups           = 0;
+            inserts           = 0;
+            lower_neighbors   = 0;
```
Comment (burgholzer) [inline include/dd/ComplexTable.hpp:359]:
```suggestion
            lowerNeighbors   = 0;
```

File: include/dd/ComplexTable.hpp [hunk 15]
```diff
@@ -334,9 +350,14 @@ namespace dd {
...
+            upper_neighbors   = 0;
```
Comment (burgholzer) [inline include/dd/ComplexTable.hpp:360]:
```suggestion
            upperNeighbors   = 0;
```

File: include/dd/ComplexTable.hpp [hunk 16]
```diff
@@ -365,6 +386,21 @@ namespace dd {
...
+        std::map<std::string, std::size_t> getStatistics() {
+            return {
+                    {"hits", hits},
+                    {"collisions", collisions},
+                    {"lookups", lookups},
+                    {"inserts", inserts},
+                    {"insert_collisions", insert_collisions},
+                    {"find_or_inserts", find_or_inserts},
+                    {"upper_neighbors", upper_neighbors},
+                    {"lower_neighbors", lower_neighbors},
```
Comment (burgholzer) [inline include/dd/ComplexTable.hpp:398]: camelCase

File: include/dd/ComplexTable.hpp [hunk 17]
```diff
@@ -380,52 +416,54 @@ namespace dd {
...
-
     private:
-        using Bucket = Entry *;
-        using Table = std::array<Bucket, NBUCKET>;
+        using Bucket = Entry*;
+        using Table  = std::array<Bucket, NBUCKET>;
 
         Table table{};
 
+        std::array<Entry*, NBUCKET> last_entry_table{};
```
Comment (burgholzer) [inline include/dd/ComplexTable.hpp:429]:
```suggestion
        std::array<Entry*, NBUCKET> lastEntryTable{};
```
what about naming this `tailTable`?

File: include/dd/ComplexTable.hpp [hunk 18]
```diff
@@ -380,52 +416,54 @@ namespace dd {
...
+
+        // table lookup statistics
+        std::size_t collisions        = 0;
+        std::size_t insert_collisions = 0;
```
Comment (burgholzer) [inline include/dd/ComplexTable.hpp:433]:
```suggestion
        std::size_t insertCollisions = 0;
```

File: include/dd/ComplexTable.hpp [hunk 19]
```diff
@@ -380,52 +416,54 @@ namespace dd {
...
+        std::size_t hits              = 0;
+        std::size_t find_or_inserts   = 0;
```
Comment (burgholzer) [inline include/dd/ComplexTable.hpp:435]:
```suggestion
        std::size_t findOrInserts   = 0;
```

File: include/dd/ComplexTable.hpp [hunk 20]
```diff
@@ -380,52 +416,54 @@ namespace dd {
...
+        std::size_t lookups           = 0;
+        std::size_t inserts           = 0;
+        std::size_t lower_neighbors   = 0;
```
Comment (burgholzer) [inline include/dd/ComplexTable.hpp:438]:
```suggestion
        std::size_t lowerNeighbors   = 0;
```

File: include/dd/ComplexTable.hpp [hunk 21]
```diff
@@ -380,52 +416,54 @@ namespace dd {
...
+        std::size_t upper_neighbors   = 0;
```
Comment (burgholzer) [inline include/dd/ComplexTable.hpp:439]:
```suggestion
        std::size_t upperNeighbors   = 0;
```

File: include/dd/ComplexTable.hpp [hunk 22]
```diff
@@ -380,52 +416,54 @@ namespace dd {
...
 
         // garbage collection
         std::size_t gcCalls = 0;
-        std::size_t gcRuns = 0;
+        std::size_t gcRuns  = 0;
         std::size_t gcLimit = 100000;
 
-        inline Entry *find_or_insert(const std::size_t key, const fp val) {
-            const fp val_tol = val - TOLERANCE;
+        inline Entry* find_or_insert(const std::size_t key, const fp val) {
```
Comment (burgholzer) [inline include/dd/ComplexTable.hpp:460]:
```suggestion
        inline Entry* findOrInsert(const std::size_t key, const fp val) {
```
is there a better name for this? I'd say `insert` itself would already be enough, but this collides with the other `insert` method.

**Review by burgholzer — CHANGES_REQUESTED** [review]
just a few minor things

File: include/dd/ComplexTable.hpp (hunk 10 above)
Comment (hillmich) [reply include/dd/ComplexTable.hpp:180]: We can even guarentee `val > p_lower->value` :)

File: include/dd/ComplexTable.hpp (hunk 11 above)
Comment (hillmich) [reply include/dd/ComplexTable.hpp:191]: Again, we can guarentee `p_upper->value > val` :)

File: include/dd/ComplexTable.hpp (hunk 16 above)
Comment (hillmich) [reply include/dd/ComplexTable.hpp:398]: Not for the keys in the map (which is supposed to offer an easy way to expose the stats via json).

File: include/dd/ComplexTable.hpp (hunk 17 above)
Comment (hillmich) [reply include/dd/ComplexTable.hpp:429]: Good suggestions!

File: include/dd/ComplexTable.hpp (hunk 16 above)
Comment (burgholzer) [reply include/dd/ComplexTable.hpp:398]: But it's just as easy in camelCase as it is in snakeCase. Why not stick to one?

File: include/dd/ComplexTable.hpp (hunk 22 above)
Comment (hillmich) [reply include/dd/ComplexTable.hpp:460]: I think the same suits the function well as it primarily tries to find elements and only inserts nothings found (alluding to short-circuit evaluation)

Comment (burgholzer) [reply include/dd/ComplexTable.hpp:460]: In C++ (and I would assume other languages as well) `std::set::insert` does precisely this.
It inserts the value if and only if it is not already found in the container, returning an iterator to the found (or inserted) element.
https://www.cplusplus.com/reference/set/set/insert/

File: include/dd/ComplexTable.hpp (hunk 16 above)
Comment (hillmich) [reply include/dd/ComplexTable.hpp:398]: okay, camelCase it is

File: include/dd/ComplexTable.hpp (hunk 22 above)
Comment (hillmich) [reply include/dd/ComplexTable.hpp:460]: On the other hand, `std::set` does not have functions to insert regardless of previous existence. I'd keep the name.

Comment (burgholzer) [reply include/dd/ComplexTable.hpp:460]: Fine 👍🏼

**Pull request merged**